src/VecSim/algorithms/hnsw/hnsw_batch_iterator.cpp

```diff
@@ -0,0 +1,206 @@
+            }
+            this->visitNode(candidate_id);
+            char *candidate_data = hnsw_index.getDataByInternalId(candidate_id);
+            float candidate_dist = dist_func(this->getQueryBlob(), (const void *)candidate_data,
+                                             space->get_data_dim());
+            this->candidates.emplace(candidate_dist, candidate_id);
+        }
+    }
+
+    // If we found fewer results than wanted, mark the search as depleted.
+    if (top_candidates.size() < hnsw_index.getEf()) {
+        this->depleted = true;
+    }
+    return top_candidates;
+}
+
+HNSW_BatchIterator::HNSW_BatchIterator(const void *query_vector, HNSWIndex *hnsw_index,
+                                       std::shared_ptr<VecSimAllocator> allocator)
+    : VecSimBatchIterator(query_vector, std::move(allocator)), index(hnsw_index), depleted(false),
+      results(this->allocator), candidates(this->allocator),
+      top_candidates_extras(this->allocator) {
+
+    this->entry_point = this->index->hnsw.getEntryPointId();
+    // Use "fresh" tag to mark nodes that were visited along the search in some iteration.
+    this->visited_list = this->index->hnsw.getVisitedList();
+    this->visited_tag = this->visited_list->getFreshTag();
+}
+
+VecSimQueryResult_List HNSW_BatchIterator::getNextResults(size_t n_res,
+                                                          VecSimQueryResult_Order order) {
+
+    // If ef_runtime lower than the number of results to return, increase it. Therefore, we assume
+    // that the number of results that return from the graph scan is at least n_res (if exist).
+    size_t orig_ef = this->index->hnsw.getEf();
+    if (orig_ef < n_res) {
+        dynamic_cast<HNSWIndex *>(this->index)->setEf(n_res);
+    }
+    auto *batch_results = array_new<VecSimQueryResult>(n_res);
+
+    // In the first iteration, we search the graph from top bottom to find the initial entry point,
+    // and then we scan the graph to get results (layer 0).
+    if (this->getResultsCount() == 0) {
+        idType bottom_layer_ep = this->index->hnsw.searchBottomLayerEP(this->getQueryBlob());
+        this->entry_point = bottom_layer_ep;
+    }
+
+    // First, take results from the "spare" results of previous iteration.
+    size_t num_results_to_add = min(this->results.size(), n_res);
+    for (int i = 0; i < num_results_to_add; i++) {
+        batch_results = array_append(batch_results, VecSimQueryResult{});
+        VecSimQueryResult_SetId(batch_results[i], this->results.top().second);
+        VecSimQueryResult_SetScore(batch_results[i], this->results.top().first);
+        this->results.pop();
+    }
+    size_t iteration_res_num = array_len(batch_results);
+
+    vecsim_stl::max_priority_queue<pair<float, idType>> top_candidates(this->allocator);
+    if (iteration_res_num < n_res) {
+        // Scan graph for more results (at least n_res, if exist), and save them.
+        top_candidates = this->scanGraph();
```

Comment:
why not assign directly?

Reply:
Because if `iteration_res_num == n_res`, than `top_candidates` will not exist (we assume that it will be empty when it is given to `prepareResults` in this case)